### PR TITLE
added sample test config files

### PR DESCRIPTION
**Pull request by mjsterckx**
**Issue number:**

N/A

**Description of changes:**

Added sample YAML files to be used with `cli`, as well as a runbook that shows how to populate and use the files.

**Testing done:**

- Populating and running `ecs-migration-test.yaml`:
```
 NAME                                TYPE              STATE              PASSED          SKIPPED          FAILED
 aws-ecs-1-initial                   Test              pass               1               0                0
 aws-ecs-2-migrate                   Test              pass               2               0                0
 aws-ecs-3-migrated                  Test              pass               1               0                0
 aws-ecs-4-migrate                   Test              pass               2               0                0
 aws-ecs-5-final                     Test              pass               1               0                0
 external-cluster                    Resource          completed
 external-cluster-instances          Resource          completed
```
- Populating and running `ecs-test.yaml`:
```
 NAME                            TYPE               STATE               PASSED           SKIPPED          FAILED
 aws-ecs                         Test               pass                1                0                0
 external-cluster                Resource           completed
 external-cluster-instances      Resource           completed
```
- Populating and running `sonobuoy-migration-test.yaml`:
```
 NAME                            TYPE               STATE               PASSED           SKIPPED          FAILED
 aws-k8s-1-initial               Test               pass                1                5772             0
 aws-k8s-2-migrate               Test               pass                2                0                0
 aws-k8s-3-migrated              Test               pass                1                5772             0
 aws-k8s-4-migrate               Test               pass                2                0                0
 aws-k8s-5-final                 Test               pass                1                5772             0
 external-cluster                Resource           completed
 external-cluster-instances      Resource           completed
```
- Populating and running `sonobuoy-test.yaml` with SONOBUOY-MODE "quick":
```
 NAME                            TYPE               STATE               PASSED           SKIPPED          FAILED
 aws-k8s                         Test               pass                1                5772             0
 external-cluster                Resource           completed
 external-cluster-instances      Resource           completed
```
- Populating and running `vmware-migration-test.yaml`:
```
 NAME                                TYPE              STATE              PASSED          SKIPPED          FAILED
 vmware-k8s-1-initial                Test              pass               1               5772             0
 vmware-k8s-2-migrate                Test              pass               2               0                0
 vmware-k8s-3-migrated               Test              pass               1               5772             0
 vmware-k8s-4-migrate                Test              pass               2               0                0
 vmware-k8s-5-final                  Test              pass               1               5772             0
 vmware-cluster                      Resource          completed
 vmware-cluster-vms                  Resource          completed
```
- Populating and running `vmware-sonobuoy-test.yaml`:
```
 NAME                            TYPE               STATE               PASSED           SKIPPED          FAILED
 vmware-k8s                      Test               pass                1                5772             0
 vmware-cluster                  Resource           completed
 vmware-cluster-vms              Resource           completed
```
**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: samples/ecs-test.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  namespace: testsys
+spec:
+  agent:
+    name: ec2-provider
+    image: <your ec2-resource-agent image URI>
+    keepRunning: true
+    configuration:
+      clusterName: ${ecs-cluster.clusterName}
+      clusterType: ecs
+      instanceCount: 2
```
Comment (stmcginnis) [inline samples/ecs-test.yaml]: I believe you will also need:

```yaml
instanceProfileArn: ${ecs-cluster.iamInstanceProfileArn}
```

File: samples/RUNBOOK.md [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+## EKS
+
+The files in `eks` are meant to be run on an EKS test cluster. You can create a new cluster using the `eksctl` tool.
+
+### ecs-migration-test
+
+```bash
+export VARIANT=
+export ARCHITECTURE=
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Lets use `aws-ecs-1` and `x86_64` as defaults.

File: samples/RUNBOOK.md [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+export ECS_TEST_AGENT_IMAGE_URI="public.ecr.aws/bottlerocket-test-system/ecs-test-agent:v0.0.5"
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: How can we make sure this aligns with the latest released version?

Comment (ecpullen) [reply samples/RUNBOOK.md]: Can we put all of the agent images together?

File: samples/RUNBOOK.md [hunk 4]
```diff
@@ -0,0 +1,353 @@
+# Bottlerocket Test System Sample Files
+
+These files contain templates that can be populated using `sed` commands and run with the `cli` tool.
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Can you add a note about `ASSUME_ROLE`?

File: samples/RUNBOOK.md [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+Examples of how to populate each one of these files can be found below. Each codeblock starts with a small number of variables that need to be populated before running the block, followed by a number of variables that contain default values (but can be modified by the user). The final `cat` command will print the populated file to stdout.
+
+## EKS
+
+The files in `eks` are meant to be run on an EKS test cluster. You can create a new cluster using the `eksctl` tool.
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Link to `eks` directory and eksctl quick start?

File: samples/RUNBOOK.md [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+cat eks/ecs-test.yaml \
+  | sed -e "s/<ECS-TEST-AGENT-IMAGE-URI>/$(echo ${ECS_TEST_AGENT_IMAGE_URI} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<ASSUME-ROLE>/$(echo ${ASSUME_ROLE} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<AWS-REGION>/$(echo ${AWS_REGION} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<ECS-RESOURCE-AGENT-IMAGE-URI>/$(echo ${ECS_RESOURCE_AGENT_IMAGE_URI} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<EC2-RESOURCE-AGENT-IMAGE-URI>/$(echo ${EC2_RESOURCE_AGENT_IMAGE_URI} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<BOTTLEROCKET-AMI-ID>/${BOTTLEROCKET_AMI_ID}/g"
+```
+
+### sonobuoy-migration-test
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Migration Testing on `aws-k8s` Variants

File: samples/RUNBOOK.md [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+
+```bash
+export VARIANT=
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Lets use `aws-k8s-1.24` and `x86_64` as a default.

File: samples/RUNBOOK.md [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+  | sed -e "s/<STARTING-VERSION>/$(echo ${STARTING_VERSION} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<METADATA-URL>/$(echo ${METADATA_URL} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<TARGETS-URL>/$(echo ${TARGETS_URL} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<MIGRATION-TEST-AGENT-IMAGE-URI>/$(echo ${MIGRATION_TEST_AGENT_IMAGE_URI} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<EKS-RESOURCE-AGENT-IMAGE-URI>/$(echo ${EKS_RESOURCE_AGENT_IMAGE_URI} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<EC2-RESOURCE-AGENT-IMAGE-URI>/$(echo ${EC2_RESOURCE_AGENT_IMAGE_URI} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<BOTTLEROCKET-AMI-ID>/${BOTTLEROCKET_AMI_ID}/g"
+```
+
+### sonobuoy-test
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Conformance Test on `aws-k8s` Variants

File: samples/RUNBOOK.md [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+  | sed -e "s/<EC2-RESOURCE-AGENT-IMAGE-URI>/$(echo ${EC2_RESOURCE_AGENT_IMAGE_URI} | sed 's/[^a-zA-Z 0-9]/\\&/g')/g" \
+  | sed -e "s/<BOTTLEROCKET-AMI-ID>/${BOTTLEROCKET_AMI_ID}/g"
+```
+
+### vmware-migration-test
+
+This codeblock assumes that your vSphere config file has been sourced. Specifically, the variables `GOVC_USERNAME`, `GOVC_PASSWORD`, `GOVC_DATACENTER`, `GOVC_DATASTORE`, `GOVC_URL`, `GOVC_NETWORK`, `GOVC_RESOURCE_POOL`, and `GOVC_FOLDER` need to be populated.
+
+```bash
+export VARIANT=
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Default to `vmware-k8s-1.24`

File: samples/RUNBOOK.md [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+## kind
+
+These files are meant to be run on a `kind` cluster.
+
+### ecs-test
+
+```bash
+export VARIANT=
+export ARCHITECTURE=
+export ACCESS_KEY_ID=
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: "access-key-id=$(aws configure get aws_access_key_id)" \
 "secret-access-key=$(aws configure get aws_secret_access_key)"

File: samples/eks/sonobuoy-migration-test.yaml [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    name: eks-provider
+    image: <EKS-RESOURCE-AGENT-IMAGE-URI>
+    keepRunning: true
+    configuration:
+      creationPolicy: ifNotExists
+      cluster_name: external-cluster
+      region: <AWS-REGION>
+      assumeRole: <ASSUME-ROLE>
+  dependsOn: []
+  destructionPolicy: never
```
Comment (ecpullen) [inline samples/eks/sonobuoy-migration-test.yaml]: Since these are samples, artifacts shouldn't be left behind.

**Review by ecpullen — CHANGES_REQUESTED** [review]
Great work. Can we make `cluster_name` configurable for all crds and have the crds names be related to the cluster name?

File: samples/RUNBOOK.md (hunk 3 above)
Comment (mjsterckx) [reply samples/RUNBOOK.md]: Aside from adding a `latest` tag, we might be able to use the `aws ecr` CLI to get the catalog from public.ecr.aws and figure out what the latest version is.

File: samples/RUNBOOK.md [hunk 12]
```diff
@@ -0,0 +1,381 @@
+# Bottlerocket Test System Sample Files
+
+These files contain templates that can be populated using `sed` commands and run with the `cli` tool.
+Examples of how to populate each one of these files can be found below. Each codeblock starts with a small number of variables that need to be populated before running the block, followed by a number of variables that contain default values (but can be modified by the user). `ASSUME_ROLE` has a default value of `~`, but you can replace this with the ARN of an AWS IAM role.
```
Comment (ecpullen) [inline samples/RUNBOOK.md]:
```suggestion
Examples of how to populate each one of these files can be found below.
Each codeblock starts with a small number of variables that need to be populated before running the block, followed by a number of variables that contain default values (but can be modified by the user).

_Note_ : `ASSUME_ROLE` has a default value of `~` (null), but you can replace this with the ARN of an AWS IAM role that should be used for all AWS calls.
```

Make sure there's only 1 sentence per line.

File: samples/RUNBOOK.md [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+The final `cat` command will print the populated file to stdout.
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Should the final command write the manifest somewhere instead?

File: samples/RUNBOOK.md [hunk 14]
```diff
@@ -0,0 +1,381 @@
...
+
+## EKS
+
+The files in [eks](./eks) are meant to be run on an EKS test cluster. You can create a new cluster using the [eksctl](https://github.com/weaveworks/eksctl/blob/main/README.md) tool.
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: `https://eksctl.io/introduction/` might be a better link for eksctl

File: samples/RUNBOOK.md [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+
+### Migration Testing on `aws-ecs` Variants
+
+```bash
+export CLUSTER_NAME=
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: Let's put a default cluster name in.

File: samples/RUNBOOK.md [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+export ARCHITECTURE="x86_64"
+export AGENT_IMAGE_VERSION=$(cli --version | sed -e "s/^.* //g")
+export ECS_TEST_AGENT_IMAGE_URI="public.ecr.aws/bottlerocket-test-system/ecs-test-agent:v${AGENT_IMAGE_VERSION}"
+export MIGRATION_TEST_AGENT_IMAGE_URI="public.ecr.aws/bottlerocket-test-system/migration-test-agent:v${AGENT_IMAGE_VERSION}"
+export ECS_RESOURCE_AGENT_IMAGE_URI="public.ecr.aws/bottlerocket-test-system/ecs-resource-agent:v${AGENT_IMAGE_VERSION}"
+export EC2_RESOURCE_AGENT_IMAGE_URI="public.ecr.aws/bottlerocket-test-system/ec2-resource-agent:v${AGENT_IMAGE_VERSION}"
+export ASSUME_ROLE="~"
+export AWS_REGION="us-west-2"
+export UPGRADE_VERSION="v1.11.1"
+export STARTING_VERSION="v1.11.0"
```
Comment (ecpullen) [inline samples/RUNBOOK.md]: I'm wondering if these should have similar logic to `AGENT_IAMGE_VERSION`

**Review by ecpullen — COMMENTED** [review]
Just a few more nits. Great work so far.

File: samples/RUNBOOK.md (hunk 16 above)
Comment (mjsterckx) [reply samples/RUNBOOK.md]: It won't be as simple as calling something like `cli --version`. Maybe a `curl` call to `public.ecr.aws/bottlerocket/<agent-image>` and figure out the latest versions that way?

Comment (ecpullen) [reply samples/RUNBOOK.md]: Let's save that for a separate pr.

**Review by ecpullen — APPROVED** [review]
LGTM 🎉

**Review by webern — COMMENTED** [review]
Are these mostly for documentation purposes?

My concern is that the chance of these being incorrect in the future is about 100%.
Can we generate them?

**Comment by ecpullen** [comment]
> Are these mostly for documentation purposes?
> 
> My concern is that the chance of these being incorrect in the future is about 100%. Can we generate them?

The samples will be used to validate the agents before a new version is released. So they will always be correct with the latest released version.

**Comment by webern** [comment]
> > Are these mostly for documentation purposes?
> > My concern is that the chance of these being incorrect in the future is about 100%. Can we generate them?
> 
> The samples will be used to validate the agents before a new version is released. So they will always be correct with the latest released version.

Is there any way to apply them to a kind cluster during CI integ tests? We know that the agents in question will not exist and the controller will report errors, but being able to apply these to an integ cluster would ensure they are not out-of-date.

**Comment by ecpullen** [comment]
> > > Are these mostly for documentation purposes?
> > > My concern is that the chance of these being incorrect in the future is about 100%. Can we generate them?
> > 
> > 
> > The samples will be used to validate the agents before a new version is released. So they will always be correct with the latest released version.
> 
> Is there any way to apply them to a kind cluster during CI integ tests? We know that the agents in question will not exist and the controller will report errors, but being able to apply these to an integ cluster would ensure they are not out-of-date.

That would just validate the crd, it wouldn't validate the individual agents configurations. A followup to this will include tests, deserializing each piece of the manifest and verifying that the configuration matches the agents configuration. (I think that's what you are asking for)

We could also add automatic generation, but I think this is a good first step.

**Comment by webern** [comment]
>That would just validate the crd, it wouldn't validate the individual agents configurations. A followup to this will include tests, deserializing each piece of the manifest and verifying that the configuration matches the agents configuration. (I think that's what you are asking for)

Sorry, yes that's even better. Unit tests that deserialize these is what I'm looking for. Any reason not to include those tests now? I'd hate to see that fall by the wayside.

**Comment by ecpullen** [comment]
> > That would just validate the crd, it wouldn't validate the individual agents configurations. A followup to this will include tests, deserializing each piece of the manifest and verifying that the configuration matches the agents configuration. (I think that's what you are asking for)
> 
> Sorry, yes that's even better. Unit tests that deserialize these is what I'm looking for. Any reason not to include those tests now? I'd hate to see that fall by the wayside.

We'd like to use these for the next release. I'm not sure how much effort the unit tests will require.

**Review by etungsten — COMMENTED** [review]
The bash snippets with all the `sed`s is quite unsightly and a bit to much for something I would be comfortable copying and running directly.

I wonder if there's a better way to go about with the templating short of just going full helm charts.

I think I would prefer using `heredoc`s more so it's cleaner. For example, 

First, change the variables like `<CLUSTER_NAME>` to `${CLUSTER_NAME}` in the template yaml file. Then...

```
### Migration Testing on `aws-ecs` Variants

CLUSTER_NAME="x86-64-aws-ecs-1"
OUTPUT_FILE="${CLUSTER_NAME}-migration.yaml"
VARIANT="aws-ecs-1"
ARCHITECTURE="x86_64"
#....

eval "cat > ${OUTPUT_FILE} <<EOF
$(<eks/ecs-migration-test.yaml)
EOF
" 2> /dev/null

```

Should give you what you want without all the `sed`s. Please try it out.

**Comment by webern** [comment]
^ rebase https://github.com/bottlerocket-os/bottlerocket-test-system/compare/541b478334f96bdb7f3e00442b2d6e3deb976e05..2a5e6d06a49ee740d106a67c736c5f31afcf6089

**Comment by webern** [comment]
^ Add a unit test that will ensure one of the sample config files stays up-to-date. https://github.com/bottlerocket-os/bottlerocket-test-system/compare/2a5e6d06a49ee740d106a67c736c5f31afcf6089..120970a8e60ff6cacafb58fe5f8c24c368042979

**Comment by webern** [comment]
> Should give you what you want without all the `sed`s. Please try it out.

We were looking at this together while writing a unit test. The problem is that we have used this syntax `${blah}` to represent substitution by the controller. In other words, using shell substitution would be extremely unwieldy (impossible?) here because we've overloaded that syntax with our own templating syntax.

**Comment by mjsterckx** [comment]
^ Added the unit tests for the remaining sample files.

**Comment by mjsterckx** [comment]
^ Replaced the many `sed` commands with a single `envsubst` call.

File: bottlerocket/samples/RUNBOOK.md [hunk 17]
```diff
@@ -0,0 +1,277 @@
+# Bottlerocket Test System Sample Files
+
+These files contain templates that can be populated using `sed` commands and run with the `cli` tool.
```
Comment (webern) [inline bottlerocket/samples/RUNBOOK.md]:
```suggestion
These files contain templates that can be populated using `envsubst` commands and run with the `cli` tool.
```

File: bottlerocket/samples/eks/ecs-migration-test.yaml [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+metadata:
+  name: ${CLUSTER_NAME}-test-1-initial
+  namespace: testsys
+spec:
+  agent:
+    name: ecs-test-agent
+    image: ${ECS_TEST_AGENT_IMAGE_URI}
+    keepRunning: true
+    configuration:
+      clusterName: ${${CLUSTER_NAME}.clusterName}
```
Comment (webern) [inline bottlerocket/samples/eks/ecs-migration-test.yaml]: escaping was not required?

**Review by webern — APPROVED** [review]

File: bottlerocket/samples/eks/ecs-migration-test.yaml (hunk 18 above)
Comment (mjsterckx) [reply bottlerocket/samples/eks/ecs-migration-test.yaml]: `envsubst` only substitutes variables that it has a value for, so escaping the outer `${` is not necessary in this case.

Comment (mjsterckx) [reply bottlerocket/samples/eks/ecs-migration-test.yaml]: Switched to an `eval` command instead so the outer `${` is now escaped.

**Comment by mjsterckx** [comment]
^ Changed `envsubst` to `eval`

**Review by etungsten — APPROVED** [review]

**Pull request merged**